### PR TITLE
[Datasets] Add telemetry for Ray Data

### PR DESCRIPTION
**Pull request by c21**
## Why are these changes needed?

This PR is to add telemetry usage tag for Ray Data. As the starting point, collect all operators being used. Given the current telemetry design (only allow put(key, value), and overwrite can happen inside GCS memory), the tag key is the logical operator, and the value is the name and count for each operator being executed in JSON string format.

For followup, a lot of useful information can be collected, such as tagging the action / consumption APIs, data size (scale), number of input files, parallelism, compute strategy (task/actor), block format, UDF batch format, etc.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Test with Ray Telemetry instruction, by launching a local cluster, and verified the tag shown up on JSON file:

```
>>> from ray.data.context import DatasetContext
>>> ctx = ray.data.context.DatasetContext.get_current()
>>> ctx.optimizer_enabled = True
>>> ds = ray.data.read_parquet("/Users/chengsu/try/parquet/data.parquet")
>>> ds = ds.map_batches(lambda x:x)
>>> ds = ds.random_shuffle()
>>> ds.fully_executed()
>>>
>>> ds = ray.data.read_parquet("/Users/chengsu/try/parquet/data.parquet")
>>> ds = ds.map_batches(lambda x:x)
>>> ds = ds.map(lambda x:x)
>>> ds = ds.filter(lambda x:True)
>>> ds.fully_executed()

(chengsu-dev) chengsu@chengs-mbp python % (chengsu-dev) chengsu@Chengs-MBP parquet % cat /tmp/ray/session_latest/usage_stats.json
{"usage_stats": {... "extra_usage_tags": {... "data_logical_ops": "{\"ReadParquetDatasource\": 2, \"MapBatches\": 2, \"RandomShuffle\": 1, \"MapRows\": 1, \"Filter\": 1}"} ...}
```

### CODE REVIEW

**Review by jjyao — APPROVED** [review]

**Comment by c21** [comment]
The failure in CI seems irrelevant, will do a rebase later. Shouldn't block review.

File: src/ray/protobuf/usage.proto [hunk 1]
```diff
@@ -94,4 +94,37 @@ enum TagKey {
   NUM_ACTOR_TASKS = 306;
   NUM_NORMAL_TASKS = 307;
   NUM_DRIVERS = 308;
+
+  // Data
+  // Read operators.
+  DATA_READ_PARQUET_OP = 400;
```
Comment (ericl) [inline src/ray/protobuf/usage.proto:100]: We shouldn't add these many tags. Can you instead put the operations inside the value? We can accumulate used operations in a comma-separated list and record on execution if it has changed.

E.g., `DATA_OPS_TAG -> {"parquet_read", "map", "flat_map"}`

**Review by ericl — CHANGES_REQUESTED** [review]
Please don't add this many tags, they will cause the table schema to get very large.

**Comment by jjyao** [comment]
> Please don't add this many tags, they will cause the table schema to get very large.

I think it's fine given extra tags are stored as map instead of struct?

File: src/ray/protobuf/usage.proto (hunk 1 above)
Comment (c21) [reply src/ray/protobuf/usage.proto:100]: I hate these many tags as well and happy to reduce as many as possible. Given current telemetry implementation, the key-value pair is buffered in GCS memory, for every one hour per cluster. So if multiple jobs run within the time window, the last job's `DATA_OPS_TAG` would overwrite the previous jobs.

Looks like this is solvable, to hold the lock of `_recorded_extra_usage_tags_lock`, and update the key `DATA_OPS_TAG` in-place to append the value (basically `_recorded_extra_usage_tags.get()` followed by `_recorded_extra_usage_tags.put()`), cc @jjyao.

**Comment by ericl** [comment]
That's definitely less bad, but will still be hard to query and analyze. It's much easier to write UDFs analyzing all data ops if they're grouped together instead of having to know the full universe of data op names.

File: src/ray/protobuf/usage.proto (hunk 1 above)
Comment (jjyao) [reply src/ray/protobuf/usage.proto:100]: I think it's fine for now to allow one job to overwrite another: statistically you still know the ops usage.

File: python/ray/data/_internal/logical/rules/record_usage_tag.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+from ray.data.datasource.json_datasource import JSONDatasource
+from ray.data.datasource.mongo_datasource import MongoDatasource
+from ray.data.datasource.numpy_datasource import NumpyDatasource
+from ray.data.datasource.parquet_base_datasource import ParquetBaseDatasource
+from ray.data.datasource.parquet_datasource import ParquetDatasource
+from ray.data.datasource.text_datasource import TextDatasource
+from ray.data.datasource.tfrecords_datasource import TFRecordDatasource
+
+
+class RecordUsageTagRule(Rule):
```
Comment (jianoaix) [inline python/ray/data/_internal/logical/rules/record_usage_tag.py:31]: How does this compare to e.g. recording usage at planning time (logical->physical translation)? That looks a bit cleaner to me.

Comment (c21) [reply python/ray/data/_internal/logical/rules/record_usage_tag.py:31]: I want to keep tracking the Dataset API usage, so think tag logical operator here (given logical operator is 1-to-1 mapping of API). If feeling awkward to put as optimization rule, I can move it before the logical optimization.

**Comment by c21** [comment]
@ericl - changed the approach to collect all operators name and count into a JSON string (see description for an example of output). Thanks.

File: python/ray/data/_internal/logical/rules/record_usage_tag.py (hunk 2 above)
Comment (c21) [reply python/ray/data/_internal/logical/rules/record_usage_tag.py]: Updated.

File: src/ray/protobuf/usage.proto (hunk 1 above)
Comment (c21) [reply src/ray/protobuf/usage.proto]: Updated to track the operator names and count in dictionary.

File: python/ray/data/_internal/logical/util.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    for child in op.input_dependencies:
+        _collect_operators_to_dict(child, ops_dict)
+
+    op_name = op.name
+    if isinstance(op, Read):
+        ds_name = type(op._datasource).__name__
+        op_name = f"Read{ds_name}"
+    elif isinstance(op, Write):
+        ds_name = type(op._datasource).__name__
+        op_name = f"Write{ds_name}"
```
Comment (c21) [inline python/ray/data/_internal/logical/util.py:40]: Will change this part accordingly after https://github.com/ray-project/ray/pull/32585/files is merged.

**Review by ericl — APPROVED** [review]

**Comment by c21** [comment]
FYI. Will wait for https://github.com/ray-project/ray/pull/32585/ before merging (for change in https://github.com/ray-project/ray/pull/32896#discussion_r1120916392).

**Review by jianoaix — APPROVED** [review]

**Comment by c21** [comment]
Checked the CI, the failure of test not relevant here. Ready to merge.

**Review by scv119 — APPROVED** [review]
you need @thomasdesr / @pcmoritz to approve the usage.proto change

File: src/ray/protobuf/usage.proto [hunk 4]
```diff
@@ -94,4 +94,9 @@ enum TagKey {
   NUM_ACTOR_TASKS = 306;
   NUM_NORMAL_TASKS = 307;
   NUM_DRIVERS = 308;
+
+  // Data
+  // Logical operators, stored in JSON format with operator name and count.
+  // Example: {"MapBatches": 2, "Filter": 1}
+  DATA_LOGICAL_OPS = 400;
```
Comment (c21) [inline src/ray/protobuf/usage.proto:101]: @pcmoritz / @thomasdesr - could you help review the change here? Thanks.

**Comment by c21** [comment]
Per offline discussion with @thomasdesr, added a whitelist of operator names, to avoid accidentially log any unwanted user-defined data.

**Review by thomasdesr — APPROVED** [review]
Boop!

**Pull request merged**